nav2_loopback_sim/src/loopback_simulator.cpp

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2018 Intel Corporation
+// Copyright (c) 2019 Samsung Research America
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include "nav2_loopback_sim/loopback_simulator.hpp"
+
+namespace nav2_loopback_sim
+{
+
+LoopbackSimulator::LoopbackSimulator()
+: Node("loopback_simulator_node")
+{
+  // Initialise subscribers
+  vel_subscriber_ = this->create_subscription<geometry_msgs::msg::Twist>(
+    "cmd_vel", 10, std::bind(&LoopbackSimulator::twistCallback, this, std::placeholders::_1));
+
+  init_pose_subscriber_ = create_subscription<geometry_msgs::msg::PoseWithCovarianceStamped>(
+    "initialpose", 10,
+    std::bind(&LoopbackSimulator::initposeCallback, this, std::placeholders::_1));
+
+  // Frequency parameter for timer function
+  double timer_frequency{0.0};
+  declare_parameter("timer_frequency", 10.0);
+  get_parameter("timer_frequency", timer_frequency);
+
+  // Create a timer
+  timer_ = create_wall_timer(
+    std::chrono::duration<double>(1.0 / timer_frequency),
+    std::bind(&LoopbackSimulator::timerCallback, this));
+
+  // Initialize tf broadcaster
+  tf_broadcaster_ = std::make_unique<tf2_ros::StaticTransformBroadcaster>(*this);
+
+  //  tf
+  tf_buffer_ = std::make_unique<tf2_ros::Buffer>(this->get_clock());
+  tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);
+}
+
+void LoopbackSimulator::twistCallback(const geometry_msgs::msg::Twist::SharedPtr msg)
+{
+  if (init_pose_set_ == false) {return;}
+
+  double dt = 0.1;  // Adjust the time step
+
+  //  Transform initpose from map to odom frame
+  if (transform_initpose_once_ == true) {
+    try {
+      geometry_msgs::msg::TransformStamped transform = tf_buffer_->lookupTransform(
+        "odom",
+        init_pose_.header.frame_id,
+        tf2::TimePointZero);
+
+      tf2::doTransform(init_pose_, odom_updated_pose_, transform);
+    } catch (const tf2::TransformException & ex) {
+      RCLCPP_ERROR(
+        rclcpp::get_logger("logger_name"),
+        "Could not transform init pose from Map to Odom.");
+    }
+
+    transform_initpose_once_ = false;
+  }
+
+  // Update the transformed pose in the local variable
+  odom_updated_pose_.header.frame_id = "odom";
+  odom_updated_pose_.header.stamp = this->now();
+
+  tf2::Quaternion base_link_orientation(
```

Comment:
check out tf2::toMsg and fromMsg functions in tf2_geometry_msgs - this will condense a bunch of this code